### PR TITLE
JDK-8322648: Improve class initialization barrier in TemplateTable::_new for PPC

### PR DESCRIPTION
**Pull request by varada1110**
ppc port implementation of https://github.com/openjdk/jdk/pull/17006

Fastdebug and Release : build and tier1 testing successful.

JBS Issue : [JDK-8322648](https://bugs.openjdk.org/browse/JDK-8322648)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322648](https://bugs.openjdk.org/browse/JDK-8322648): Improve class initialization barrier in TemplateTable::_new for PPC (**Enhancement** - P4)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17518/head:pull/17518` \
`$ git checkout pull/17518`

Update a local copy of the PR: \
`$ git checkout pull/17518` \
`$ git pull https://git.openjdk.org/jdk.git pull/17518/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17518`

View PR using the GUI difftool: \
`$ git pr show -t 17518`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17518.diff">https://git.openjdk.org/jdk/pull/17518.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17518#issuecomment-1903909025)

### CODE REVIEW

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
You need to adapt the succeeding code and remove the dependent `crnand` instruction. I suggest to use `cmpdi(CCR0, Rscratch, InstanceKlass::fully_initialized);`.

**Comment by varada1110** [comment]
> You need to adapt the succeeding code and remove the dependent `crnand` instruction. I suggest to use `cmpdi(CCR0, Rscratch, InstanceKlass::fully_initialized);`.

Thank you @TheRealMDoerr for the review. I have applied the suggested changes.

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
Sorry, my comment was wrong. "fully_initialized" check is done in `clinit_barrier`. We only need
```
    __ lwz(Rinstance_size, in_bytes(Klass::layout_helper_offset()), RinstanceKlass);
    __ andi_(R0, Rinstance_size, Klass::_lh_instance_slow_path_bit);
    __ bne(CCR0, Lslow_case);
```
(See other platforms for reference.)
Bugs in this code are likely to cause interpreter performance loss and no test failures. The whole purpose of this change is to improve interpreter performance. So, this should be checked. We should either check the petclinic benchmark referenced in the original issue or at least compare performance using -Xint.

File: src/hotspot/cpu/ppc/templateTable_ppc_64.cpp [hunk 1]
```diff
@@ -3803,15 +3803,18 @@ void TemplateTable::_new() {
     __ sldi(Roffset, Rindex, LogBytesPerWord);
     __ load_resolved_klass_at_offset(Rcpool, Roffset, RinstanceKlass);
 
-    // Make sure klass is fully initialized and get instance_size.
-    __ lbz(Rscratch, in_bytes(InstanceKlass::init_state_offset()), RinstanceKlass);
+    // Make sure klass is initialized.
+    assert(VM_Version::supports_fast_class_init_checks(), "Optimization requires support for fast class initialization checks");
+    __ clinit_barrier(Rcpool, R16_thread, nullptr /*L_fast_path*/, &Lslow_case);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/templateTable_ppc_64.cpp:3808]: `Rcpool` is the wrong parameter. You need to use `RinstanceKlass`.

Comment (varada1110) [reply src/hotspot/cpu/ppc/templateTable_ppc_64.cpp:3808]: Thank you for the correction @TheRealMDoerr . I have applied the code change. Tier1 test is running. I will push once it is done.

Comment (varada1110) [reply src/hotspot/cpu/ppc/templateTable_ppc_64.cpp]: build successful and no failures in tier1

**Comment by TheRealMDoerr** [comment]
Seems like you have missed my previous comment: https://github.com/openjdk/jdk/pull/17518#pullrequestreview-1847895107
I don't think it's correct without that. And please remove the comment "// get instance_size.". It doesn't tell anything which is not obvious.
We should definitely measure performance with -Xint. Maybe even "time java -Xint -version" shows a difference between correct and incorrect code.

**Comment by varada1110** [comment]
> Seems like you have missed my previous comment: [#17518 (review)](https://github.com/openjdk/jdk/pull/17518#pullrequestreview-1847895107) I don't think it's correct without that. And please remove the comment "// get instance_size.". It doesn't tell anything which is not obvious. We should definitely measure performance with -Xint. Maybe even "time java -Xint -version" shows a difference between correct and incorrect code.

I see a minor difference in real time and CPU time

Without the code change : 
````
real    0m0.714s
user    0m0.381s
sys     0m0.026s
````

With the code change : 
````
real    0m0.711s
user    0m0.380s
sys     0m0.026s`
```

**Comment by TheRealMDoerr** [comment]
Ok, the difference is probably below noise. Maybe we can try jvm98 with -Xint. I think the patch should look like this:
```
diff --git a/src/hotspot/cpu/ppc/templateTable_ppc_64.cpp b/src/hotspot/cpu/ppc/templateTable_ppc_64.cpp
index 84ecfc4f934..74bea6ac9ac 100644
--- a/src/hotspot/cpu/ppc/templateTable_ppc_64.cpp
+++ b/src/hotspot/cpu/ppc/templateTable_ppc_64.cpp
@@ -3803,16 +3803,15 @@ void TemplateTable::_new() {
     __ sldi(Roffset, Rindex, LogBytesPerWord);
     __ load_resolved_klass_at_offset(Rcpool, Roffset, RinstanceKlass);
 
-    // Make sure klass is fully initialized and get instance_size.
-    __ lbz(Rscratch, in_bytes(InstanceKlass::init_state_offset()), RinstanceKlass);
+    // Make sure klass is initialized.
+    assert(VM_Version::supports_fast_class_init_checks(), "Optimization requires support for fast class initialization checks");
+    __ clinit_barrier(RinstanceKlass, R16_thread, nullptr /*L_fast_path*/, &Lslow_case);
+
     __ lwz(Rinstance_size, in_bytes(Klass::layout_helper_offset()), RinstanceKlass);
 
-    __ cmpdi(CCR1, Rscratch, InstanceKlass::fully_initialized);
     // Make sure klass does not have has_finalizer, or is abstract, or interface or java/lang/Class.
     __ andi_(R0, Rinstance_size, Klass::_lh_instance_slow_path_bit); // slow path bit equals 0?
-
-    __ crnand(CCR0, Assembler::equal, CCR1, Assembler::equal); // slow path bit set or not fully initialized?
-    __ beq(CCR0, Lslow_case);
+    __ bne(CCR0, Lslow_case);
 
     // --------------------------------------------------------------------------
     // Fast case:
```

**Comment by TheRealMDoerr** [comment]
I can run it, but please update your PR first (as shown above). CCR1 is used uninitialized in your current version.

**Comment by varada1110** [comment]
Thank you Martin. I have applied the changes

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the update! JVM98 performance results with -Xint -XX:-ProfileInterpreter look good (no regression). Some sub-benchmarks seem to be a very little faster, but there's no big difference.

**Comment by varada1110** [comment]
Thank you Martin for running the benchmark and reviewing the code

/integrate

**Comment by TheRealMDoerr** [comment]
We usually wait for a 2nd review, but this PR is tiny and we have run a lot of tests. So, let's ship it!
/sponsor

**Pull request closed without merging**